### PR TITLE
sp/1.9.1-new-tutorial-image-processing

### PR DESCRIPTION
**Pull request by 5675sp**

### CODE REVIEW

File: Driverless AI/image-processing-in-driverless-ai/image-processing-in-driverless-ai.md [hunk 1]
```diff
@@ -0,0 +1,595 @@
...
+
+In the case that our dataset is not similar to the ImageNet dataset or we want to improve the results of our model using ImageNet architectures, we can use fine-tuning. 
+
+When enabling fine-tuning, we are not limited to retrain only the classifier section of the CNN, but we are also able to retrain the feature extraction stage: the convolutional and pooling layers. 
+
+**Note**: In practice, networks are fine-tuned when trained on a large dataset like the ImageNet. In other words, with fine-tuning, we continue the training of the architecture with the smaller dataset we have imported(running back-propagation). Fine-tuning will only work well if the smaller dataset is not so different from the original dataset (ImageNet) our architecture was trained. Once again, the pre-trained model will contain learned features relevant to our classification problem.
```
Comment (ybabakhin) [inline Driverless AI/image-processing-in-driverless-ai/image-processing-in-driverless-ai.md:304]: I would exclude this sentence or tell that it will work better (instead of only work well)
`Fine-tuning will only work well if the smaller dataset is not so different from the original dataset (ImageNet) our architecture was trained.`

As, in practice, fine-tuning works for any dataset no matter how large it differs from the ImageNet. And the reason for that is that pre-trained models in the early layers learnt some simple representations like edges, strokes, etc. And these representation are always more useful than start training from the random weights.

And here we might say `classification or regression`:
`Once again, the pre-trained model will contain learned features relevant to our classification problem. `

File: Driverless AI/image-processing-in-driverless-ai/image-processing-in-driverless-ai.md [hunk 2]
```diff
@@ -0,0 +1,595 @@
...
+
+When enabling fine-tuning, we are not limited to retrain only the classifier section of the CNN, but we are also able to retrain the feature extraction stage: the convolutional and pooling layers. 
+
+**Note**: In practice, networks are fine-tuned when trained on a large dataset like the ImageNet. In other words, with fine-tuning, we continue the training of the architecture with the smaller dataset we have imported(running back-propagation). Fine-tuning will only work well if the smaller dataset is not so different from the original dataset (ImageNet) our architecture was trained. Once again, the pre-trained model will contain learned features relevant to our classification problem. 
+
+Before we explore a rerun of the first experiment from task one, let us end this task by mentioning one more default setting that was enabled by default during the first experiment. 
+
+Every time we define a classification learning problem with a feature-vector, we are creating a feature space. Consequently, Driverless AI allows you to enable the dimensionality of the feature (embeddings) space by Image Transformer. The following are options that you can choose from:
```
Comment (ybabakhin) [inline Driverless AI/image-processing-in-driverless-ai/image-processing-in-driverless-ai.md:308]: I don't understand this sentence:
`Every time we define a classification learning problem with a feature-vector, we are creating a feature space`
Could you, please, elaborate on it?

We take the embeddings vector from the last global average pooling layer of the pre-trained model. For different models it has different dimensionalities (usually ranging from 512 to 2k). And in DAI we suppress these representations to lower number in order to suppress the number of features generated by the transformer.

File: Driverless AI/image-processing-in-driverless-ai/image-processing-in-driverless-ai.md [hunk 3]
```diff
@@ -0,0 +1,595 @@
...
+
+![list-of-augmentations-for-fine-tuning-used-for-image-transformer](assets/list-of-augmentations-for-fine-tuning-used-for-image-transformer.png)
+
+**NOTE**: Augmentations for fine-tuning used for the Image Transformer is only available when Fine-tuning is **enabled**.  
+
+As well, when fine-tuning is enabled, you can specify the number of epochs for fine-tuning **ImageNet pre-trained** models used for the **Image Transformer**. This value defaults to **2**. This default setting can be found and change in the **IMAGE** tab inside the **EXPERT SETTINGS**.
+
+Now that you know how to rerun the experiment with fine-tuning let's explore the new experiment (Embeddings-Transformer-With-Fine-Tuning). 
+
+![finish-experiment-car_deals_with_fine_tuning](assets/finish-experiment-car_deals_with_fine_tuning.png)
```
Comment (ybabakhin) [inline Driverless AI/image-processing-in-driverless-ai/image-processing-in-driverless-ai.md:371]: It seems that this experiment hasn't used Image features at all in the final model. As it has only 50 features and there is no ImageVectorizer features in the variable importance.

I would try to re-run it with different settings/architecture/target variable in order to see image features here.

File: Driverless AI/image-processing-in-driverless-ai/image-processing-in-driverless-ai.md [hunk 4]
```diff
@@ -0,0 +1,595 @@
...
+
+For our second experiment, we see that the validation score for the final pipeline is RMSE = 4114.0424 +/- 231.9882. Recall that the RMSE for the first experiment was 4058.833 +/- 154.7854. Enabling fine-tuning didn't improve the RMSE; instead, it did the opposite. The RMSE increase by ~55.209.
```
Comment (ybabakhin) [inline Driverless AI/image-processing-in-driverless-ai/image-processing-in-driverless-ai.md:373]: Maybe the first experiment also hasn't used images in the final model. So, this comparison and conclusions afterwards could not be applied here.
I would rather change settings/dataset/etc. in order to show that Fine-Tuning helps improve the model quality.

File: Driverless AI/image-processing-in-driverless-ai/image-processing-in-driverless-ai.md [hunk 5]
```diff
@@ -0,0 +1,595 @@
...
+
+For the most part, fine-tuning will lead to better results, though there are times when that will not be the case. Performance (an improvement on a scorer) depends on the type of problem you have: If: 
+
+- Our dataset(car_deals_train) is smaller and similar to the original one(ImageNet) - you need to be careful with fine-tuning because it could be the case that other learning models can achieve better results(e.g., SVM). In this case, fine-tuning is not necessary for our car deals dataset. 
+
+- When the new dataset is larger and similar to the original, having more data will not over-fit the model. Therefore, we can say with confidence that fine-tuning can achieve better results. 
+
+In our case, the Xception model has been trained on ~1.2 million training images with another 50,000 images for validation and 100,000 images for testing. Our car_deals_train dataset with 26k images will be considered small. That is why fine-tuning did not improve the RMSE because it could be the case that an SVM or a linear classifier can improve the scorer.
```
Comment (ybabakhin) [inline Driverless AI/image-processing-in-driverless-ai/image-processing-in-driverless-ai.md:381]: 26k images dataset is quite large for the current Computer Vision state. But finetuning could still not help much, as the ImageNet already has been trained on a lot of various car models. And I would also not mention SVM, as we're not using it anywhere and it's quite an old approach.

File: Driverless AI/image-processing-in-driverless-ai/image-processing-in-driverless-ai.md [hunk 6]
```diff
@@ -0,0 +1,595 @@
...
+        - **Without fine-tuning**: the experiment will usually finish faster but has the lowest performance 
+        - **With fine-tuning**: the experiment will be a bit slower, but should produce better results  
+        - **Automatic Image Model**: the slowest by far, but produces the best results 
+- When your dataset image column is crucial to your regression or classification problem, it is best to use the second approach: Automatic Image Model. Hence, if images are not playing a crucial role in your experiment, you can use the Embeddings Transformer.
+- **Python scoring** and **C++ MOJO Scoring** are supported for the image transformer.
+- Presently, only **Python scoring** is supported for **Automatic Image Model**
+
+With the above in mind, you are ready to generate your Image Models. Note: as of now, Driverless AI supports the following problem types: 
+
+- Embeddings Transformer
```
Comment (ybabakhin) [inline Driverless AI/image-processing-in-driverless-ai/image-processing-in-driverless-ai.md:559]: We don't support `Embeddings Transformer` problem type. We just support the usual DAI's classification (single-label) and regression.

**Review by ybabakhin — CHANGES_REQUESTED** [review]
Thank you very much for the tutorial @5675sp! It looks great, I've added a couple of minor comments.

The only major one is: did the image transformer experiments actually use image features in the final model? You could find more details in my comments.

Also, should we provide links to the initial datasets on Kaggle? Maybe I've missed it, but just in case here are the sources:
* Car Deals dataset: https://www.kaggle.com/gogotchuri/myautogecardetails
 * Histopathology images: https://www.kaggle.com/c/histopathologic-cancer-detection/data

File: Driverless AI/image-processing-in-driverless-ai/image-processing-in-driverless-ai.md [hunk 7]
```diff
@@ -0,0 +1,595 @@
...
+
+As mentioned above, we can define a neural network architecture by choosing an existing ImageNet architecture, but how can we avoid the need to train our neural network from scratch? Usually, neural networks are initialized with random weights that reach a level of value that allows the network to classify the image input after a series of epochs are executed. With the just mentioned, the question that must be asked now is what if we could initialize those weights to certain values that we know beforehand are already good to classify a certain dataset. In our case, the car deals dataset. If the weights are predefined to correct values, we will not need to wait for a good number of epochs, and therefore, the weights will have it much more manageable. And the just mentioned above is achieved through transfer learning. Besides transfer learning, this can also be achieved with fine-tuning.
```
[the remark this answers is not in the document]
Comment (5675sp) [reply Driverless AI/image-processing-in-driverless-ai/image-processing-in-driverless-ai.md:298]: @ybabakhin in this case, when fine-tuning is not enabled, transfer learning is applied under which method? (when fine-tuning is enabled transfer learning is applied through the fine-tuning method in other words transfer learning and fine-tuning are counterpart?)

File: Driverless AI/image-processing-in-driverless-ai/image-processing-in-driverless-ai.md (hunk 2 above)
Comment (5675sp) [reply Driverless AI/image-processing-in-driverless-ai/image-processing-in-driverless-ai.md:311]: With your comments in mind, it has been rewritten as follows: 

Driverless AI allows you to enable the dimensionality of the feature (embeddings) space by Image Transformer. We take the embeddings vector from the last global average pooling layer of the pre-trained model. For different models, it has different dimensionalities (usually ranging from 512 to 2k). And in DAI, we suppress these representations to a lower number to suppress the transformer's number of features. The following are options that you can choose from: 

Does that ^make sense now?
